File: app/src/main/java/me/devsaki/hentoid/fragments/library/UpdateSuccessDialogFragment.java
package me.devsaki.hentoid.fragments.library;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.RecyclerView;

import com.mikepenz.fastadapter.FastAdapter;
import com.mikepenz.fastadapter.adapters.ItemAdapter;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import me.devsaki.hentoid.R;
import me.devsaki.hentoid.json.GithubRelease;
import me.devsaki.hentoid.retrofit.GithubServer;
import me.devsaki.hentoid.viewholders.GitHubReleaseItem;
import timber.log.Timber;

import static androidx.core.view.ViewCompat.requireViewById;

/**
 * Launcher dialog for the library refresh feature
 */
public class UpdateSuccessDialogFragment extends DialogFragment {

    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
    private final ItemAdapter<GitHubReleaseItem> itemAdapter = new ItemAdapter<>();

    public static void invoke(FragmentManager fragmentManager) {
        UpdateSuccessDialogFragment fragment = new UpdateSuccessDialogFragment();
        fragment.show(fragmentManager, "usdf");
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedState) {
        View rootView = inflater.inflate(R.layout.dialog_library_update_success, container, false);

        FastAdapter<GitHubReleaseItem> releaseItemAdapter = FastAdapter.with(itemAdapter);
        RecyclerView releaseItem = requireViewById(rootView, R.id.changelogReleaseItem);
        releaseItem.setAdapter(releaseItemAdapter);

        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        getReleases();
    }

    @Override
    public void onDestroy() {
        compositeDisposable.clear();
        super.onDestroy();
    }

    private void getReleases() {
        compositeDisposable.add(GithubServer.API.getLatestRelease() // No need to run on BG thread; retrofit already makes async calls
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onCheckSuccess, this::onCheckError)
        );
    }

    private void onCheckSuccess(GithubRelease latestReleaseInfo) {
        itemAdapter.add(new GitHubReleaseItem(latestReleaseInfo));
    }

    private void onCheckError(Throwable t) {
        Timber.w(t, "Error fetching GitHub latest release data");
    }
}

File: app/src/main/java/me/devsaki/hentoid/fragments/library/UpdateSuccessDialogFragment.kt
package me.devsaki.hentoid.fragments.library

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.lifecycleScope
import com.mikepenz.fastadapter.FastAdapter
import com.mikepenz.fastadapter.adapters.ItemAdapter
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import me.devsaki.hentoid.databinding.DialogLibraryUpdateSuccessBinding
import me.devsaki.hentoid.json.GithubRelease
import me.devsaki.hentoid.retrofit.GithubServer
import me.devsaki.hentoid.viewholders.GitHubReleaseItem
import timber.log.Timber

/**
 * "update success" dialog
 */
class UpdateSuccessDialogFragment : DialogFragment() {

    // UI
    private var _binding: DialogLibraryUpdateSuccessBinding? = null
    private val binding get() = _binding!!

    // === VARIABLES
    private val itemAdapter: ItemAdapter<GitHubReleaseItem> = ItemAdapter()


    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedState: Bundle?
    ): View {
        _binding = DialogLibraryUpdateSuccessBinding.inflate(inflater, container, false)

        val releaseItemAdapter = FastAdapter.with(itemAdapter)
        binding.recyclerView.adapter = releaseItemAdapter

        return binding.root
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }

    override fun onViewCreated(rootView: View, savedInstanceState: Bundle?) {
        super.onViewCreated(rootView, savedInstanceState)
        getReleases()
    }

    private fun getReleases() {
        lifecycleScope.launch {
            var response: GithubRelease? = null
            withContext(Dispatchers.IO) {
                try {
                    response = GithubServer.api.latestRelease.execute().body()
                } catch (e: Exception) {
                    onCheckError(e)
                }
            }
            response.let {
                if (null == it) Timber.w("Error fetching GitHub latest release data (empty response)")
                else onCheckSuccess(it)
            }
        }
    }

    private fun onCheckSuccess(latestReleaseInfo: GithubRelease) {
        itemAdapter.add(GitHubReleaseItem(latestReleaseInfo))
    }

    private fun onCheckError(t: Throwable) {
        Timber.w(t, "Error fetching GitHub latest release data")
    }

    companion object {
        fun invoke(parent: Fragment) {
            val fragment = UpdateSuccessDialogFragment()
            fragment.show(parent.childFragmentManager, null)
        }

        fun invoke(parent: FragmentActivity) {
            val fragment = UpdateSuccessDialogFragment()
            fragment.show(parent.supportFragmentManager, null)
        }
    }
}